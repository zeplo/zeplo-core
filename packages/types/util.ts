
export type Maybe<T> = T | null

export type JsonValue =  boolean | number | string | null | JsonArray | JsonMap
export interface JsonMap { [key: string]: JsonValue }
export interface JsonArray extends Array<JsonValue> {}

export type DeepPartial<T> = T extends Function ? T : (T extends object ? { [P in keyof T]?: DeepPartial<T[P]>; } : T);

/**
 * Get all of the keys to which U can be assigned.
 */
export type OnlyKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? K : never
}[keyof T];

/**
 * Get the interface containing only properties to which U can be assigned.
 */
export type OnlyUndefined<T> = {
  [K in OnlyKeys<T, undefined>]: T[K]
}

/**
 * Get all of the keys except those to which U can be assigned.
 */
export type ExcludeKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? never : K
}[keyof T];

/**
 * Get the interface containing no properties to which U can be assigned.
 */
export type ExcludeUndefined<T> = {
  [K in ExcludeKeys<T, undefined>]: T[K]
}

/**
 * Get the interface where all properties are optional.
 */
export type Optional<T> = {[K in keyof T]?: T[K]};

/**
 * Get the interface where properties that can be assigned undefined are
 * also optional.
 */
export type UndefinedOptional<T> = ExcludeUndefined<T> & Optional<OnlyUndefined<T>>

export type DeepUndefinedOptional<T> = T extends (...args: any[]) => any
  ? T
  : T extends any[]
  ? _DeepRequiredArray<T[number]>
  : T extends object
  ? _DeepRequiredObject<UndefinedOptional<T>>
  : T;

 export interface _DeepRequiredArray<T>
  extends Array<DeepUndefinedOptional<T>> {}

  export type _DeepRequiredObject<T> = {
  [P in keyof T]: DeepUndefinedOptional<T[P]>
}