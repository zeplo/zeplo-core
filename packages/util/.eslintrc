{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "parser":  "@typescript-eslint/parser",
  "env": {
    "browser": true,
    "node": true,
    "es6": true,
    "mocha": true,
    "jest": true
  },
  "rules": {
    "semi": ["error", "never"],
    "no-underscore-dangle": 0,
    "space-before-function-paren": ["error", "always"],
    "arrow-body-style": 0,
    "no-use-before-define": ["error", { "functions": false }],
    "arrow-parens": 0,
  }
}
