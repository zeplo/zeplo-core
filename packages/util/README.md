# @zeplo/util

Utility functions that are used across packages.