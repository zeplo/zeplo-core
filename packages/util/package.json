{
  "name": "@zeplo/util",
  "version": "2.1.0",
  "main": "dist/index.js",
  "files": [
    "dist/*"
  ],
  "license": "MIT",
  "scripts": {
    "clean": "rimraf dist",
    "dev": "env-cmd ts-node",
    "build": "tsc",
    "prepublish": "yarn clean && yarn build"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/fs-extra": "^9.0.13",
    "@types/lodash": "^4.14.178",
    "@types/node": "^17.0.18",
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "env-cmd": "^10.1.0",
    "eslint": "^8.9.0",
    "rimraf": "^3.0.2",
    "ts-node": "^10.5.0",
    "ts-node-dev": "^1.1.8",
    "typescript": "^4.5.5"
  },
  "dependencies": {
    "@zeplo/errors": "^2.0.8",
    "@zeplo/types": "^2.1.0",
    "express": "^4.17.3",
    "express-async-handler": "^1.2.0",
    "lodash": "^4.17.21",
    "uuid": "^8.3.2"
  }
}
