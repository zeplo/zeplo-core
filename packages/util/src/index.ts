export { cleanQuery, cleanHeaders } from './clean'
export { 
  parseRequest, 
  parseBody, 
  getBackoff, 
  getStartTimeFromFeatures,
  getRawFeatures,
  parseRawRequestObject, 
  createRequestJob ,
  getValueFromQueryOrHeader,
} from './parse'
export { 
  createStepsJobs,
  hasCyclicalDependency,
} from './steps'
export { 
  requestHandler,
  jsonRequestHandler,
  queryToString,
  queryToNumber,
  parseQuery,
} from './request'