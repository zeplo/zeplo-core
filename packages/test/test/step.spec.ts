/**
 * @jest-environment node
 */

import { omit } from 'lodash'
import { createWorkspace } from '#/util/data'
import waitFor from '#/util/waitFor'
import queue from '#/util/queue'

jest.setTimeout(60000)

describe('step.spec', () => {
  const workspace = 'test-workspace-steps'
  const token = workspace

  let cleanup: () => Promise<void>
  beforeAll(async () => {
    cleanup = await createWorkspace(workspace, token)
  })
  afterAll(() => cleanup())

  test('single step request', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
    }])

    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const children2 = await queue.get(`/requests?_token=${token}&filters={"trace": "${id}" }`)
    expect(children2.data.length).toBe(1)
    expect(children2.data[0]).toMatchObject({
      status: 'SUCCESS',
      source: 'STEP',
      trace: id,
    })
  })

  test('two step request with dependency', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post?_step=A',
    }, {
      url: 'https://postman-echo.com/post?_requires=A',
    }])

    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const children2 = await queue.get(`/requests?_token=${token}&filters={"trace": "${id}" }`)
    expect(children2.data.length).toBe(2)
    expect(children2.data[0]).toMatchObject({
      status: 'SUCCESS',
      source: 'STEP',
      trace: id,
    })
    expect(children2.data[1]).toMatchObject({
      status: 'SUCCESS',
      source: 'STEP',
      trace: id,
    })
  })

  test('multi step request with double dependency', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post?_step=A',
    }, {
      url: 'https://postman-echo.com/post?_requires=A',
    }, {
      url: 'https://postman-echo.com/post?_requires=A&_step=B',
    }, {
      url: 'https://postman-echo.com/post?_requires=A&_step=C',
    }, {
      url: 'https://postman-echo.com/post?_requires=C&_step=D',
    }, {
      url: 'https://postman-echo.com/post?_requires=D,B',
    }])

    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const children2 = await queue.get(`/requests?_token=${token}&filters={"trace": "${id}" }`)
    expect(children2.data.length).toBe(6)
    expect(children2.data[0]).toMatchObject({
      status: 'SUCCESS',
      source: 'STEP',
      trace: id,
    })
    expect(children2.data[1]).toMatchObject({
      status: 'SUCCESS',
      source: 'STEP',
      trace: id,
    })
  })

  test('error all steps have requires', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post?_requires=B',
    }]).catch((err) => err.response)
    expect(res.status).toBe(400)
    expect(res.data).toMatchObject({
      error: { message: 'At least one step must have no requires parameter, as that is the starting step' },
    })
  })

  test('error missing required step name', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post?_step=A',
    }, {
      url: 'https://postman-echo.com/post?_requires=B',
    }]).catch((err) => err.response)
    expect(res.status).toBe(400)
    expect(res.data).toMatchObject({
      error: { message: 'Step is required by another step, but not provided' },
    })
  })

  test('error cylical dependencies - step calls itself', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
    }, {
      url: 'https://postman-echo.com/post?_step=A&_requires=A',
    }]).catch((err) => err.response)
    expect(res.status).toBe(400)
    expect(res.data).toMatchObject({
      error: { message: 'Cyclical dependencies in steps (e.g. a -> b -> a)' },
    })
  })

  test('error cylical dependencies - multi step cycle', async () => {
    const res = await queue.post(`/step?_token=${token}`, [{
      url: 'https://postman-echo.com/post?_step=0',
    }, {
      url: 'https://postman-echo.com/post?_step=A&_requires=B,0',
    }, {
      url: 'https://postman-echo.com/post?_step=B&_requires=C',
    }, {
      url: 'https://postman-echo.com/post?_step=C&_requires=A',
    }]).catch((err) => err.response)
    expect(res.status).toBe(400)
    expect(res.data).toMatchObject({
      error: { message: 'Cyclical dependencies in steps (e.g. a -> b -> a)' },
    })
  })
})
