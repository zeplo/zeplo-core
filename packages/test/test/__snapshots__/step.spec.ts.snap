// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step.spec multi step request with double dependency 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/json",
    "user-agent": "axios/0.26.0",
  },
  "host": "step",
  "method": "POST",
  "params": Object {},
  "path": "/",
  "scheme": "https",
  "url": "https://step",
}
`;

exports[`step.spec single step request 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/json",
    "user-agent": "axios/0.26.0",
  },
  "host": "step",
  "method": "POST",
  "params": Object {},
  "path": "/",
  "scheme": "https",
  "url": "https://step",
}
`;

exports[`step.spec two step request with dependency 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/json",
    "user-agent": "axios/0.26.0",
  },
  "host": "step",
  "method": "POST",
  "params": Object {},
  "path": "/",
  "scheme": "https",
  "url": "https://step",
}
`;
